File: cpp/digit_problem/Maximal Rectangle.cpp
class Solution {
public:
    int getmax(vector<int> &line)
    {
        stack<int> num;
        num.push(0);
        int iret = 0;
        
        for(int i = 1;i < line.size();i++)
        {
            if(line[i] < line[num.top()])
            {
                while(num.size() != 0 && line[i] < line[num.top()])
                {
                    int cur = num.top();
                    num.pop();
                    
                    if(iret < (i - num.top() - 1) * line[cur])
                        iret = (i - num.top() - 1) * line[cur];
                }
            }
            
            num.push(i);
        }
        
        return iret;
    }
    int maximalRectangle(vector<vector<char> > &matrix) {
        
        if(matrix.size() == 0)
            return 0;
        if(matrix[0].size() == 0)
            return 0;
            
        int iret = 0;
        vector<int> currentline(matrix[0].size()+2,0);
        
        for(int i = 0;i < matrix.size();i++)
        {
            for(int j = 0; j < matrix[i].size();j++)
            {
                if(matrix[i][j] == '0')
                {
                    currentline[j+1] = 0;
                }
                else if(matrix[i][j] == '1')
                {
                    currentline[j+1] += 1;
                }
            }
            
            int tmp = getmax(currentline);
            if(tmp > iret)
                iret = tmp;
        }
        
        return iret;
        
    }
};

File: py/Maximal Rectangle.py
class Solution:
    # @param matrix, a list of lists of 1 length string
    # @return an integer
    def max(self,height):
        stack = []
        
        height.insert(0,0)
        height.append(0)
        
        top = -1
        iret = 0
        
        for i in range(len(height)):
            if top == -1:
                stack.append(i)
                top += 1
            if height[stack[top]] <= height[i]:
                stack.append(i)
                top += 1
            else:
                while top != -1 and height[stack[top]] > height[i]:
                    icur = stack.pop()
                    top -= 1
                    if (i - stack[top] - 1) * height[icur] > iret:
                        iret = (i - stack[top] - 1) * height[icur]
                        
                stack.append(i)
                top += 1
        
        return iret
                
        
    def maximalRectangle(self, matrix):
        dp = []
        if matrix == []:
            return 0
            
        for i in range(len(matrix)):
            dp.append([0]*len(matrix[0]))    
            
        for i in range(len(matrix[0])):
            dp[0][i] = int(matrix[0][i])
            
        for i in range(1,len(matrix)):
            for j in range(len(matrix[0])):
                if int(matrix[i][j]) == 1:
                    dp[i][j] = dp[i-1][j] + 1
        iret = 0
        
        for i in range(len(matrix)):
            tmp =  self.max(dp[i]) 
            if tmp > iret:
                iret = tmp
        
        return iret